solr/core/src/java/org/apache/solr/core/backup/repository/BackupRepository.java

```diff
@@ -142,6 +142,7 @@ default String getBackupLocation(String override) {
    */
   OutputStream createOutput(URI path) throws IOException;
 
+  // TODO define whether this should also create any nonexistent parent directories. (i.e. is this 'mkdir', or 'mkdir -p')
```

Comment:
> I'm not sure if there are use cases where we don't want to create parent directories if needed.

Good question. There's no obvious use case in the backup code afaik.  Though I could see an argument that **not** creating parent directories is a safer API, as it'd throw an error if a consumer misspelled a dirname, etc.  (Not saying that's how I'd lean, just thinking aloud.)

Anyway, I created a JIRA here per your suggestion: https://issues.apache.org/jira/browse/SOLR-15359

Reply:
To further complicate the issue, it appears something changed from Solr 8.x to 9.x -- now, Solr _expects_ it to be `mkdir -p` behavior when taking backups. As I'm cleaning up our S3 code and moving it to 9.x repo, this behavior change broke our code (b/c ours throws exceptions if the parent dir does not exist).

Reply:
Huh, that could've been from the incremental-backup support that I added a few months back, but that wasn't my intention.  And it's been a bit since I ran this test, but I was nearly certain that LocalFileSystemRepository would error-out if a parent directory was missing.

But if that's what you're seeing, it's hard to argue with.  Any chance you still have a stacktrace handy that shows where that error cropped up in Solr's backup code?  (Might be worth continuing this discussion on SOLR-15089 since JIRA's a little more durable/find-able than these PRs.  But I'll try to remember to check back both places )

Reply:
Oh -- the stacktrace I was referring to is thrown from our code. Basically, we check if the parent dir exists when creating a dir, and throw an exception if it's not there.

If I send this backup cmd to 8.2 and 9.0
`/solr/admin/collections?action=BACKUP&repository=blob&location=blob:/foo&collection=$COLLECTION&name=$BACKUP_NAME`

In 8.2, all the files are under $BACKUP_NAME, but in 9.0 all files are under $BACKUP_NAME/$COLLECTION.

Reply:
I realize I misread your ask for the stacktrace -- you're looking for where in Solr code someone's asking for a subdir and assuming `mkdir -p` behavior. Here it is:

```
2021-04-29 21:49:16.472 INFO  (OverseerThreadFactory-15-thread-2-processing-n:localhost:8983_solr) [c:collection1619732944   ] o.a.s.c.a.c.BackupCmd Starting to backup ZK data for backupName=backup1619732944
2021-04-29 21:49:16.481 ERROR (OverseerThreadFactory-15-thread-2-processing-n:localhost:8983_solr) [c:collection1619732944   ] o.a.s.c.a.c.OverseerCollectionMessageHandler Collection: collection1619732944 operation: backup failed:java.io.IOException: org.apache.solr.blob.client.BlobException: Parent directory doesn't exist, path=foo/backup1619732944/collection1619732944/zk_backup_0/
        at org.apache.solr.blob.backup.BlobBackupRepository.createDirectory(BlobBackupRepository.java:120)
        at org.apache.solr.core.backup.BackupManager.downloadConfigDir(BackupManager.java:249)
        at org.apache.solr.cloud.api.collections.BackupCmd.call(BackupCmd.java:124)
        at org.apache.solr.cloud.api.collections.OverseerCollectionMessageHandler.processMessage(OverseerCollectionMessageHandler.java:117)
        at org.apache.solr.cloud.OverseerTaskProcessor$Runner.run(OverseerTaskProcessor.java:531)
        at org.apache.solr.common.util.ExecutorUtil$MDCAwareThreadPoolExecutor.lambda$execute$0(ExecutorUtil.java:226)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

And you're right, makes more sense to continue the discussion on SOLR-15359, since this PR is closed now. I'll copy this comment over there.